examples/websockets/src/main.rs

```diff
@@ -59,58 +76,169 @@ async fn main() {
+    // since each client gets individual statemachine. Thus, waiting for this
+    // client to finish getting his greetings does not prevent other clients form
+    // connecting to server.
+    for i in 1..5 {
         if socket
-            .send(Message::Text(String::from("Hi!")))
+            .send(Message::Text(String::from(format!("Hi {} times!", i))))
             .await
             .is_err()
         {
-            println!("client disconnected");
+            println!("client abruptly disconnected");
             return;
         }
-        tokio::time::sleep(std::time::Duration::from_secs(3)).await;
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+    }
+
+    // By splitting socket we can send and receive at the same time. In this example we will send
+    // unsolicited messages to client based on some sort of server's internal event (i.e .timer).
+    let (mut sender, mut receiver) = socket.split();
+
+    // Spawn a task that will push 100 messages to the client (does not matter what client does)
+    let mut send_task = tokio::spawn(async move {
+        let mut i = 20;
+        while i > 0 {
+            // In any websocket error, break loop.
+            if sender
+                .send(Message::Text(format!("{} messages left...", i)))
+                .await
+                .is_err()
+            {
+                break;
+            }
+
+            tokio::time::sleep(std::time::Duration::from_millis(300)).await;
+            i -= 1;
+        }
+        if i == 0 {
+            println!("Sending close to {}...", who);
+            if let Err(e) = sender
+                .send(Message::Close(Some(CloseFrame {
+                    code: 1000,
+                    reason: Cow::from("Goodbye"),
+                })))
+                .await
+            {
+                println!("Could not send Close due to {}, probably it is ok?", e);
+            }
+        }
+        i
+    });
+
+    // This second task will receive messages from client and print them on server console
+    let mut recv_task = tokio::spawn(async move {
+        let mut cnt = 0;
+        while let Some(Ok(msg)) = receiver.next().await {
+            cnt += 1;
+            // print message and break if instructed to do so
+            if process_message(msg, who) {
+                break;
+            }
+        }
+        cnt
+    });
+
+    // If any one of the tasks exit, abort the other.
+    tokio::select! {
+        rv_a = (&mut send_task) => {
+            match rv_a {
+                Ok(a)=>println!("{} messages remaining to send", a),
+                Err(a)=>println!("Error sending messages {:?}", a)
+            }
+            recv_task.abort();
+        },
+        rv_b = (&mut recv_task) =>{
+            match rv_b {
+                Ok(b)=>println!("Received {} messages", b),
+                Err(b)=>println!("Error receiving messages {:?}", b)
+            }
+            send_task.abort();
+        }
+    }
```

Comment:
So odd formatting going on here. Sadly rustfmt doesn't like macros 😕

Reply:
I've noticed a few odd things about it too. At the very least it is consistently odd.

Reply:
This feels in the spirit for what rustfmt would do

```rust
    tokio::select! {
        rv_a = (&mut send_task) => {
            match rv_a {
                Ok(a) => println!("{} messages remaining to send", a),
                Err(a) => println!("Error sending messages {:?}", a)
            }
            recv_task.abort();
        },
        rv_b = (&mut recv_task) => {
            match rv_b {
                Ok(b) => println!("Received {} messages", b),
                Err(b) => println!("Error receiving messages {:?}", b)
            }
            send_task.abort();
        }
    }
```